File: LC/[0399] Evaluate Division - chatGpt.cpp
class Solution {
    unordered_map<string, vector<pair<string, double>>> Map;
public:
    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        Map.clear();
        for(int ii=0; ii<equations.size(); ++ii){
            Map[equations[ii][0]].push_back({equations[ii][1], values[ii]}); //a->[(b, v1),(c, v2), ...]
            Map[equations[ii][1]].push_back({equations[ii][0], 1.0/values[ii]});
        }           
        vector<double> results;
        for(int ii=0; ii<queries.size(); ++ii){
            if(Map.find(queries[ii][0])==Map.end() || Map.find(queries[ii][1])==Map.end()){
                results.push_back(-1);
                continue;
            }
            unordered_set<string> visited({queries[ii][0]});
            results.push_back(DFS(queries[ii][0], queries[ii][1], visited));
        }
        return results;
    }

    double DFS(string & A, string & B, unordered_set<string>&visited){

        if(A==B) return 1.0;
        for(int ii=0; ii<Map[A].size(); ++ii){
            string C = Map[A][ii].first;
            //if(B==C) return Map[A][ii].second;
            if(visited.find(C)!=visited.end()) continue;
            visited.insert(C);
            double val1 = Map[A][ii].second;
            double val2 = DFS(C, B, visited);
            //visited.erase(C); --> this is not required.
            if(val2!=-1.0) return val1*val2;
        }
        return -1.0;
    }
};

File: LC/[0399] Evaluate Division - chatGpt.java
class Solution {
    private static class Edge {
        String variable;
        double weight;

        Edge(String variable, double weight) {
            this.variable = variable;
            this.weight = weight;
        }
    }

    private final Map<String, List<Edge>> graph = new HashMap<>();

    public double[] calcEquation(
        List<List<String>> equations,
        double[] values,
        List<List<String>> queries
    ) {
        graph.clear();

        // Build the bidirectional weighted graph.
        for (int i = 0; i < equations.size(); i++) {
            String a = equations.get(i).get(0);
            String b = equations.get(i).get(1);

            // a / b = values[i]
            graph.computeIfAbsent(a, key -> new ArrayList<>())
                 .add(new Edge(b, values[i]));

            // b / a = 1 / values[i]
            graph.computeIfAbsent(b, key -> new ArrayList<>())
                 .add(new Edge(a, 1.0 / values[i]));
        }

        double[] results = new double[queries.size()];

        for (int i = 0; i < queries.size(); i++) {
            String source = queries.get(i).get(0);
            String target = queries.get(i).get(1);

            if (!graph.containsKey(source) ||
                !graph.containsKey(target)) {
                results[i] = -1.0;
                continue;
            }

            Set<String> visited = new HashSet<>();
            visited.add(source);

            results[i] = dfs(source, target, visited);
        }

        return results;
    }

    private double dfs(
        String current,
        String target,
        Set<String> visited
    ) {
        if (current.equals(target)) {
            return 1.0;
        }

        for (Edge edge : graph.get(current)) {
            String next = edge.variable;

            if (visited.contains(next)) {
                continue;
            }

            visited.add(next);

            double remainingValue = dfs(next, target, visited);

            if (remainingValue != -1.0) {
                return edge.weight * remainingValue;
            }
        }

        return -1.0;
    }
}